File: frontend/src/components/benchmark/content/counting_sort/implementations/counting_sort.java
public class CountingSort {
    private int findMax(int[] arr) {
        int maxVal = arr[0];
        for (int val : arr) {
            if (val > maxVal) maxVal = val;
        }
        return maxVal;
    }

    private int findMin(int[] arr) {
        int minVal = arr[0];
        for (int val : arr) {
            if (val < minVal) minVal = val;
        }
        return minVal;
    }

    public void sort(int[] arr) {
        if (arr == null || arr.length == 0) return;

        int maxVal = findMax(arr);
        int minVal = findMin(arr);
        int range = maxVal - minVal + 1;

        int[] count = new int[range];
        int[] output = new int[arr.length];

        for (int val : arr) {
            count[val - minVal]++;
        }

        for (int i = 1; i < count.length; i++) {
            count[i] += count[i - 1];
        }

        for (int i = arr.length - 1; i >= 0; i--) {
            output[count[arr[i] - minVal] - 1] = arr[i];
            count[arr[i] - minVal]--;
        }

        for (int i = 0; i < arr.length; i++) {
            arr[i] = output[i];
        }
    }

    public void sortStable(int[] arr) {
        sort(arr);
    }

    public void sortDescending(int[] arr) {
        sort(arr);
        reverse(arr);
    }

    private void reverse(int[] arr) {
        int i = 0, j = arr.length - 1;
        while (i < j) {
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
            i++;
            j--;
        }
    }
}

File: frontend/src/components/benchmark/content/counting_sort/implementations/counting_sort.rs
pub struct CountingSort;

impl CountingSort {
    fn find_max(arr: &[i32]) -> i32 {
        let mut max_val = arr[0];
        for &val in arr {
            if val > max_val {
                max_val = val;
            }
        }
        max_val
    }

    fn find_min(arr: &[i32]) -> i32 {
        let mut min_val = arr[0];
        for &val in arr {
            if val < min_val {
                min_val = val;
            }
        }
        min_val
    }

    pub fn sort(arr: &mut [i32]) {
        if arr.is_empty() {
            return;
        }

        let max_val = Self::find_max(arr);
        let min_val = Self::find_min(arr);
        let range = (max_val - min_val + 1) as usize;

        let mut count = vec![0; range];
        let mut output = vec![0; arr.len()];

        for &val in arr.iter() {
            count[(val - min_val) as usize] += 1;
        }

        for i in 1..count.len() {
            count[i] += count[i - 1];
        }

        for i in (0..arr.len()).rev() {
            let idx = (arr[i] - min_val) as usize;
            output[count[idx] - 1] = arr[i];
            count[idx] -= 1;
        }

        arr.copy_from_slice(&output);
    }

    pub fn sort_stable(arr: &mut [i32]) {
        Self::sort(arr);
    }

    pub fn sort_descending(arr: &mut [i32]) {
        Self::sort(arr);
        arr.reverse();
    }
}